File: src/api/private/discussion-post-command.ts
import { RequestFN } from "~/core/request-function";
import { DiscussionCommand, EntityState, TabLocation, type SessionHandle } from "~/models";

export const discussionPostCommand = async (session: SessionHandle, command: DiscussionCommand | "", extra: any): Promise<void> => {
  let payload;

  switch (command) {
    case DiscussionCommand.brouillon:
      payload = {
        commande: command,
        brouillon: typeof extra.id === "number" ? {
          E: EntityState.CREATION,
          N: extra.id
        } : {
          E: EntityState.MODIFICATION,
          N: extra.id
        },

        contenu: extra.content,

        messagePourReponse: {
          G: 0,
          N: extra.replyMessageID
        },

        listeDestinataires: [],
        listeFichiers: [],
        objet: ""
      };
      break;
    case "":
      payload = {
        commande: command,
        bouton: {
          N: 0,
          G: extra.button
        },

        brouillon: {
          N: extra.id
        },

        contenu: extra.content,
        listeDestinataires: [],
        listeFichiers: [],

        messagePourReponse: {
          G: 0,
          N: extra.replyMessageID
        }
      };
      break;
    default:
      payload = {
        commande: command,
        listePossessionsMessages: extra.possessions
      };
  }

  const request = new RequestFN(session, "SaisieMessage", {
    _Signature_: { onglet: TabLocation.Discussions },
    donnees: payload
  });

  await request.send();
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/private/discussion-post-command.kt
package ink.literate.pawnote.api.private

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.models.DiscussionCommand
import ink.literate.pawnote.models.EntityState
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun discussionPostCommand(
    session: SessionHandle,
    command: DiscussionCommand? = null,
    extra: JsonObject
) {
  val payload: JsonObject

  when (command) {
    DiscussionCommand.brouillon -> payload = buildJsonObject {
          put("commande", command.value)
          put(
              "brouillon",
              when (extra["id"]!!.jsonPrimitive.intOrNull != null) {
                true ->
                    buildJsonObject {
                      put("E", EntityState.CREATION.code)
                      put("N", extra["id"]!!)
                    }
                false ->
                    buildJsonObject {
                      put("E", EntityState.MODIFICATION.code)
                      put("N", extra["id"]!!)
                    }
              })

          put("contenu", extra["content"]!!)

          putJsonObject("messagePourReponse") {
            put("G", 0)
            put("N", extra["replyMessageID"]!!)
          }

          putJsonArray("listeDestinataires") {}
          putJsonArray("listeFichiers") {}
          put("objet", "")
        }
    null -> payload = buildJsonObject {
          put("commande", "")
          putJsonObject("bouton") {
            put("N", 0)
            put("G", extra["button"]!!)
          }

          putJsonObject("brouillon") { put("N", extra["id"]!!) }

          put("contenu", extra["content"]!!)

          putJsonArray("listeDestinataires") {}
          putJsonArray("listeFichiers") {}

          putJsonObject("messagePourReponse") {
            put("G", 0)
            put("N", extra["replyMessageID"]!!)
          }
        }
    else -> payload = buildJsonObject {
          put("commande", command.value)
          put("listePossessionsMessages", extra["possessions"]!!)
        }
  }

  val request =
      RequestFN(
          session.information,
          "SaisieMessage",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Discussions.code) }

                put("donnees", payload)
              }))

  request.send()
}
